### PR TITLE
Net core 3 support & targeting cleanup

### PR DESCRIPTION
**Pull request by ig-sinicyn**
What's done so far:
1. Support for core 3 added.
2. No name/type resolution conflicts on legacy frameworks. Targeting code switched to Theraot.Core shims where possible. Some extension methods (`.ToHashSet()`, `.dictionary.GetValueOrDefault()`) moved to its own namespace `CodeJam.Collections.Backported` as there is no other way to avoid name conflicts with extension methods defined in other packages.
3. All features work on all frameworks. Partially done, see below.
4. Conditional build statements are cleaned up. What's supported and what's not is much more clearer now.

On feature availability. All places that cannot be backported to legacy frameworks are marked with `PUBLIC_API_CHANGES` comment. They fall into two major categories:
1. Types are not included in the current release of Theraot.Core and we do not want to backport them by ourselves (`DynamicExpression` as example). As a result entire Mapping stack (Mapping, `TypeAccessor`, `ComparerBuilder` etc) is not available under net35 and earlier.
2. Entire functionality is missing. As example, `.GetMetadataAttribute()` methods are not available under `netstandard 1.3` as the target does not include `ICustomAttributeProvider` type.

### CODE REVIEW

File: Build/BuildScripts/CodeJam.AppVeyor.NUnit.Tests.ps1 [hunk 1]
```diff
@@ -1,22 +1,21 @@
...
-$includePerfTests = "*.Tests.Performance.dll"
-$exclude = "Experimental\\.*?\\CodeJam.Tests.Performance.dll"
 
 mkdir "$env:APPVEYOR_BUILD_FOLDER\_Results" -ErrorAction SilentlyContinue
 
 $wc = New-Object System.Net.WebClient
 
-$targetsDotNet = "net47","net46","net45","net40","net35","net20"
+#run .net tests
+$targetsDotNet = "net472","net461","net45","net40","net35","net20"
```
Comment (NN---) [inline Build/BuildScripts/CodeJam.AppVeyor.NUnit.Tests.ps1]: Should we add net48, net471, net470 too ?

Comment (ig-sinicyn) [reply Build/BuildScripts/CodeJam.AppVeyor.NUnit.Tests.ps1]: Done.

File: Build/BuildScripts/CodeJam.PerfTests.AppVeyor.NUnit.Tests.ps1 [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+#if ($LastExitCode -ne 0) { $host.SetShouldExit($LastExitCode) }
+#echo "UploadFile: https://ci.appveyor.com/api/testresults/nunit3/$env:APPVEYOR_JOB_ID from $logFileName"
+#$wc.UploadFile("https://ci.appveyor.com/api/testresults/nunit3/$env:APPVEYOR_JOB_ID", "$logFileName")
+#if ($LastExitCode -ne 0) {
+#	echo "FAIL: UploadFile: https://ci.appveyor.com/api/testresults/nunit3/$env:APPVEYOR_JOB_ID from $logFileName"
+#	$host.SetShouldExit($LastExitCode)
+#}
+
+#run .net core tests
+$targetsDotNetCore = "netcoreapp2.0","netcoreapp1.1","netcoreapp1.0"
```
Comment (NN---) [inline Build/BuildScripts/CodeJam.PerfTests.AppVeyor.NUnit.Tests.ps1]:
```suggestion
$targetsDotNetCore = "netcoreapp3.0, netcoreapp2.0","netcoreapp1.1","netcoreapp1.0"
```

Comment (ig-sinicyn) [reply Build/BuildScripts/CodeJam.PerfTests.AppVeyor.NUnit.Tests.ps1]: Perftests are disabled for now and I do not want to change related scripts until they will be ready to use.

Comment (NN---) [reply Build/BuildScripts/CodeJam.PerfTests.AppVeyor.NUnit.Tests.ps1]: Ok

File: Build/BuildScripts/CodeJam.AppVeyor.NUnit.Tests.ps1 [hunk 3]
```diff
@@ -1,22 +1,21 @@
...
-$includePerfTests = "*.Tests.Performance.dll"
-$exclude = "Experimental\\.*?\\CodeJam.Tests.Performance.dll"
 
 mkdir "$env:APPVEYOR_BUILD_FOLDER\_Results" -ErrorAction SilentlyContinue
 
 $wc = New-Object System.Net.WebClient
 
-$targetsDotNet = "net47","net46","net45","net40","net35","net20"
+#run .net tests
+$targetsDotNet = "net48","net472","net471","net47","net461","net45","net40","net35","net20"
```
Comment (NN---) [inline Build/BuildScripts/CodeJam.AppVeyor.NUnit.Tests.ps1]: Do we need a separate net46 ?

Comment (ig-sinicyn) [reply Build/BuildScripts/CodeJam.AppVeyor.NUnit.Tests.ps1]: I'd prefer to keep the list as short as possible. Let's keep the current one and add more frameworks as new targeting issue arrive.

**Comment by ig-sinicyn** [comment]
@NN--- looks like there's no critical issues with the PR. May I merge it?

**Comment by NN---** [comment]
Seems good.
Maybe instead of TODO we should have an issue in GitHub ?
TODO in code is usually not actionable and will be forgotten.

And update NuGet :)

**Comment by ig-sinicyn** [comment]
> Seems good.
> Maybe instead of TODO we should have an issue in GitHub ?
> TODO in code is usually not actionable and will be forgotten.
> 
> And update NuGet :)


> TODO in code is usually not actionable and will be forgotten.
Yep, I'll add an issue into theraot core and then will reference to it from СJ issue.

**Pull request merged**